stan/math/rev/fun/quad_form.hpp

```diff
@@ -93,6 +94,65 @@ class quad_form_vari : public vari {
 
   quad_form_vari_alloc<Ta, Ra, Ca, Tb, Rb, Cb>* impl_;
 };
+
+/**
+ * Return the quadratic form \f$ B^T A B \f$.
+ *
+ * Symmetry of the resulting matrix is not guaranteed due to numerical
+ * precision.
+ *
+ * @tparam Mat1 type of the first (square) matrix
+ * @tparam Mat2 type of the second matrix
+ *
+ * @param A square matrix
+ * @param B second matrix
+ * @param symmetric indicates whether the output should be made symmetric
+ * @return The quadratic form
+ * @throws std::invalid_argument if A is not square, or if A cannot be
+ * multiplied by B
+ */
+template <typename Mat1, typename Mat2,
+          require_all_matrix_t<Mat1, Mat2>* = nullptr,
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
+inline auto quad_form_impl(const Mat1& A, const Mat2& B, bool symmetric) {
+  check_square("quad_form", "A", A);
+  check_multiplicable("quad_form", "A", A, "B", B);
+
+  using return_t
+      = promote_var_matrix_t<decltype(value_of(B).transpose().eval()
+                                      * value_of(A) * value_of(B).eval()),
+                             Mat1, Mat2>;
+
+  auto arena_A = to_arena(to_var_value_if<!is_constant<Mat1>::value>(A));
+  auto arena_B = to_arena(to_var_value_if<!is_constant<Mat2>::value>(B));
```

Comment:
If it is a `mat<var>`, then it needs converted to a `var<mat>`, otherwise pass through.

The other option from these `to_var_value_if` statements are more function definitions and requires. Yes `to_var_value_if` and friends are confusing, but I do not think this is more confusing than:

```
require_eigen_st<is_var, Mat1>* = nullptr,
require_var_matrix_t<Mat2>* = nullptr
```

which is something like you'd need to add to the functions to specify that one of the things is an Eigen-like Matrix with a var and one of the other things is a var matrix.

The result of the `to_var_value_if`, regardless of what it is, needs to end up in the arena cause we use it in the reverse pass (hence everything is wrapped in a `to_arena`).

Reply:
:clap: Pull :clap: Out :clap: The :clap: Control :clap: Flow :clap:

All of these problems go away if we just pull the control flow out into the function. There's a lot of good reasons to prefer pulling out control flow

1. Each path from the start to the return is linear. This makes it a lot easier to read and I don't have to go through each line and think, "Well what if X is var and y is not or x is var and y is var and....". In a block of the control flow I'm aware of which is which.

2. Removes the need for widgets like `forward_as<>()` and more type trait layers like `to_*_if<>()`. All the paths do different things and work on different types. Yinz keep saying this leads to code duplication but I disagree! The code is different in each path, it makes the code more _verbose_ which, imo, is very different than duplication.

3. Uses `value_of()` less (which requires making a `holder<>` class for `Matrix<var>`) 

4. Adding more additional control flow is less annoying. Like if we want to do the `.noalias()` optimization for `var<mat>` like 

```cpp
      if (is_var_matrix<Mat1>::value) {
        arena_A.adj().noalias() += arena_B * (res.adj() * arena_B.transpose());
      } else {
        arena_A.adj() += arena_B * (res.adj() * arena_B.transpose());
      }
```

When we bunch the control flow together with the code this turns into

```cpp
      if (!is_constant<Mat1>::value) {
        if (is_var_matrix<Mat1>) {
          forward_as<arena_t<promote_scalar_t<var, Mat1>>>(arena_A).adj().noalias()
              += value_of(arena_B) * C_adj_B_t;
        } else {
          forward_as<arena_t<promote_scalar_t<var, Mat1>>>(arena_A).adj()
              += value_of(arena_B) * C_adj_B_t;
        }
      }
```

That's a mouthful!

I have an example of breaking out the control flow for `quad_form` below. imo, way more clear! Yeah there's more literal lines of code, but like that's fine to me because I think it's simpler and easier to read.

https://github.com/stan-dev/math/compare/feature/quad_form...feature/quad_form-pullout-logic

Reply:
If you have like 3 inputs or something then yeah I think it's a good idea to use the `to_*_if()` stuff cuz otherwise you end up with like a bunch of paths which feels silly, but for binary functions I think pulling out the control flow is much cleaner

Reply:
And also the one downside is that for some reason the version the pulls out the flow has to use `val_op()` otherwise we run into `const` correctness errors due to something internal in Eigen

Reply:
> All of these problems go away if we just pull the control flow
> "Well what if X is var and y is not or x is var and y is var and...."
> Yinz keep saying this leads to code duplication but I disagree!

Sure. I disagree. I don't think any of this code is easy, and I have to read it very carefully line by line anyway.

I can pull out the control flow though (though I don't think it gets rid of the `to_var_value` problem).

> Uses value_of() less (which requires making a holder<> class for Matrix<var>)

I was wondering why you were swapping those all the time. I just liked `value_of` more cause I don't have to worry about the types. If we add a plain type overload could `value_of` just point to `.val()` (or `.val_op()`) instead of using a holder?

> Like if we want to do the .noalias() optimization for var<mat> like

Well if we converted `mat<var>` to `var<mat>` we wouldn't need the extra logic. But if it's necessary sure. `.noalias()` is good optimization.

> If you have like 3 inputs or something

For what it's worth, it is COming Soon (TM) with `trace_gen_quad_form`, but we can do different code there.

Reply:
> Sure. I disagree. I don't think any of this code is easy, and I have to read it very carefully line by line anyway.

The asm etc that's generated for the two versions can be v different. Like for example with `to_arena_if<>()`, even when the `bool` template is `false` it's not a no-op and still has to do make an Eigen map. For smaller N stuff like that matters

> I can pull out the control flow though (though I don't think it gets rid of the to_var_value problem).

For this I'd use something like the version I put up where it doesn't do the conversion

> I was wondering why you were swapping those all the time. I just liked value_of more cause I don't have to worry about the types. 

Yeah so `value_of()` for Eigen types calls

```cpp
template <typename EigMat, require_eigen_t<EigMat>* = nullptr,
          require_not_st_arithmetic<EigMat>* = nullptr>
inline auto value_of(EigMat&& M) {
  return make_holder(
      [](auto& a) {
        return a.unaryExpr([](const auto& scal) { return value_of(scal); });
      },
      std::forward<EigMat>(M));
}
```

So it's good to call `value_of()` of when we are doing stuff like

```cpp
// A is an input argument
arena_t<promote_scalar_t<double, Mat1>> arena_A(value_of(A));
```

But once we have the `arena_t` we should just call `.val()` etc.

> If we add a plain type overload could value_of just point to .val() (or .val_op()) instead of using a holder?

Actually @t4c1 why is there a reason we don't have a separate version for plain and non-plain Eigen types? I think we could have an overload for when it's an Eigen matrix of vars that are not an expression and just return `.val()`

> Well if we converted mat<var> to var<mat> we wouldn't need the extra logic. But if it's necessary sure. .noalias() is good optimization.

Yeah it would be v cool if we could do that swap and have it be efficient but idt it's worth the cost of swapping stuff around.

> For what it's worth, it is COming Soon (TM) with trace_gen_quad_form, but we can do different code there.

Oh yeah, I'd be a monster if I asked for breaking up the control flow for anything more than a binary function

Reply:
> Uses value_of() less (which requires making a holder<> class for Matrix)

> If we add a plain type overload could value_of just point to .val() (or .val_op()) instead of using a holder?

Holder is only needed for rvalue inputs (which would not work without it). For lvalues `make_holder` just calls the given functor with given argument.

> Actually @t4c1 why is there a reason we don't have a separate version for plain and non-plain Eigen types? I think we could have an overload for when it's an Eigen matrix of vars that are not an expression and just return .val()

`.val()` does exactly the same thing as `value_of` called on Matrices.

Reply:
> .val() does exactly the same thing as value_of called on Matrices.

The slight difference is that ```value_of``` will always return a ```CWiseUnaryOp``` whereas ```.val()``` will return either a ```CWiseUnaryView``` or ```CWiseUnaryOp```, depending on const-ness